### PR TITLE
Fix type assert generation on MacOS

### PR DESCRIPTION
**Pull request by isinyaaa**
This makes type asserts in Python instead of bash, hopefully making it more portable.

## Description
Performing a `touch api/openapi/model-registry.yaml && make` currently won't work on MacOS, after #322. To fix this, we should drop the `-P` flag on `grep` as it's GNU specific.
As suggested by @tarilabs, the script could be significantly simplified by using a Python script to gather type asserts deterministically.

## How Has This Been Tested?
Ran various variations of the `make` command to ensure it won't break.

## Merge criteria:

- All the commits have been [_signed-off_](https://github.com/kubeflow/community/tree/master/dco-signoff-hook#signing-off-commits)  (To pass the `DCO` check)

- [ ] The commits have meaningful messages; the author will squash them after approval or in case of manual merges will ask to merge with squash.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work.
- [ ] Code changes follow the [kubeflow contribution guidelines](https://www.kubeflow.org/docs/about/contributing/).

If you have UI changes

- [ ] The developer has added tests or explained why testing cannot be added.
- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Verify that UI/UX changes conform the UX guidelines for Kubeflow.

cc: @tarilabs @dhirajsb

### CODE REVIEW

File: scripts/gen_type_asserts.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        with path.open() as f:
+            yield "".join(f.readlines()[14:])
+        path.unlink()
+
+
+if __name__ == "__main__":
+    if len(argv) > 1:
+        root = Path(argv[1])
+        assert root.is_dir(), f"{root} is not a directory"
+    else:
```
Comment (isinyaaa) [inline scripts/gen_type_asserts.py:18]: on a second thought we might not even need this

File: scripts/gen_type_asserts.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            # skip first 14 lines of boilerplate
+            for raw in islice(f.readlines(), 14, None):
+                line = raw.rstrip()
+                buf.append(line)
+                if line.startswith("func"):
+                    in_func = True
+                elif line.startswith("}") and in_func:
+                    in_func = False
+                    yield "\n".join(buf)
+                    buf.clear()
```
Comment (tarilabs) [inline scripts/gen_type_asserts.py:22]: I'd take the approach of re-ordering in the file by keeping the header, as shown here: https://github.com/kubeflow/model-registry/pull/404/files#diff-30505a65ba1ad63e91b0b363704b01d852f43625afea32db6355c7626679a7aaR8-R28

this way we don't have to rely on a fixed `14` my2c

**Review by tarilabs — COMMENTED** [review]
what I lot about this approach is that it also take in charge of processing the model_ files

**Review by tarilabs — APPROVED** [review]
- https://github.com/kubeflow/model-registry/pull/403#pullrequestreview-2316680793
- https://github.com/kubeflow/model-registry/pull/404#issuecomment-2362926052

/lgtm
/approve

**Pull request merged**